File: components/kotlin/app/src/main/kotlin/stlc/REPL.kt
package stlc

import stlc.bci.compileTo
import java.io.File
import kotlin.system.exitProcess

fun main(args: Array<String>) {
    if (args.isEmpty()) {
        println("Welcome to the REPL of the Lambda Calculus Interpreter!")
        println("Type \".quit\" to exit.")
        println("Enter a multi-line expression with ;; as a terminator.")

        while (true) {
            val input = readline().trim()

            if (input == ".quit") {
                println("bye")
                break
            }

            try {
                executeInput(input)
            } catch (e: LanguageException) {
                println(e.formatMessage())
            }
        }
    } else if (args.size == 1) {
        val input = File(args[0]).readText()
        try {
            executeInput(input)
        } catch (e: LanguageException) {
            println(e.formatMessage())
            exitProcess(1)
        }
    } else if (args.size == 2) {
        println("Compiling ${args[0]} to ${args[1]}")
        try {
            compileTo(File(args[0]).readText(), args[1])
        } catch (e: LanguageException) {
            println(e.formatMessage())
            exitProcess(1)
        }
    } else {
        println("Usage: tlca [file-name] [output-file]")
    }
}

private fun renameTypeVariables(t: Type): Type {
    var i = 0

    fun nextVar(): String =
        if (i < 26)
            ('a' + i++).toString()
        else
            "t${i++ - 26}"

    val vars = t.ftv().toList()
    val subst = Subst(vars.zip(vars.map { TVar(nextVar()) }).toMap())
    return t.apply(subst)
}

private fun executeInput(input: String) {
    val (value, type) = execute(input)

    if (type is TArr) {
        println("function: ${renameTypeVariables(type)}")
    } else {
        println("$value: ${renameTypeVariables(type)}")
    }
}

private fun readline(): String {
    var result = ""

    while (true) {
        if (result.isEmpty())
            print("> ")
        else
            print(". ")

        val s = readln()
        result = result + "\n" + s.trimEnd()

        if (result.endsWith(";;"))
            return result.dropLast(2)
    }
}

File: components/deno/Repl.ts
import { execute as executeProgram, formatError } from "./Interpreter.ts";
import { Subst, TVar, Type } from "./Typing.ts";

const readline = (): string | null => {
  let result = "";

  while (true) {
    const line = prompt(result === "" ? ">" : ".");

    if (line === null) {
      return null;
    }

    result = (result + "\n" + line).trim();

    if (result.endsWith(";;")) {
      return result.substring(0, result.length - 2);
    }
  }
};

const renameTypeVariables = (type: Type): Type => {
  let i = 0;

  const nextVar = (): string => {
    if (i < 26) {
      return String.fromCharCode(97 + i++);
    } else {
      return `t${i++}`;
    }
  };

  const vars = type.ftv();
  const subst = new Subst(
    new Map([...vars].map((v) => [v, new TVar(nextVar())])),
  );
  return type.apply(subst);
};

const execute = (line: string) => {
  const [value, type] = executeProgram(line);

  console.log(`${value}: ${renameTypeVariables(type)}`);
};

if (Deno.args.length === 0) {
  console.log("Welcome to the REPL of the Lambda Calculus Interpreter!");
  console.log('Type ".quit" to exit.');
  console.log("Enter a multi-line expression with ;; as a terminator.");

  while (true) {
    const line = readline();

    if (line == null) {
      break;
    }

    if (line.trim() === ".quit") {
      console.log("bye...");
      Deno.exit(0);
    } else {
      try {
        execute(line);
      } catch (e) {
        console.log(formatError(e));
      }
    }
  }
} else if (Deno.args.length === 1) {
  try {
    execute(Deno.readTextFileSync(Deno.args[0]));
  } catch (e) {
    console.log(formatError(e));
  }
} else {
  console.error("Invalid arguments");
}
